cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt

```diff
@@ -0,0 +1,241 @@
+     * Used for performing both write-time and read-time based cache expiration
+     * as well as size-based eviction.
+     *
+     * Note that a write is also considered an access.
+     */
+    private val accessQueue: MutableSet<CacheEntry<Key, Value>>?
+
+    /**
+     * Whether to perform size based evictions.
+     */
+    private val evictsBySize = maxSize >= 0
+
+    /**
+     * Whether to perform write-time based expiration.
+     */
+    private val expiresAfterWrite = expireAfterWriteNanos > 0
+
+    /**
+     * Whether to perform access-time (both read and write) based expiration.
+     */
+    private val expiresAfterAccess = expireAfterAccessNanos > 0
+
+    init {
+        // writeQueue is required if write expiry is enabled
+        writeQueue = takeIf { expiresAfterWrite }?.let {
+            Collections.synchronizedSet(ReorderingLinkedHashSet())
+        }
+
+        // accessQueue is required if either read expiry is enabled or size-based eviction is enabled
+        accessQueue = takeIf { expiresAfterAccess || evictsBySize }?.let {
+            Collections.synchronizedSet(ReorderingLinkedHashSet())
+        }
+    }
+
+    override fun get(key: Key): Value? {
+        val nowNanos = clock.currentTimeNanos
+        return cacheEntries[key]?.let {
+            if (it.isExpired(nowNanos)) {
+                // clean up expired entries and return null
+                expireEntries(nowNanos)
+                null
+            } else {
+                // update eviction metadata and return the value
+                it.recordRead(nowNanos)
+                it.value
+            }
+        }
+    }
+
+    override fun get(key: Key, loader: () -> Value): Value {
+        synchronized(this) {
+            val nowNanos = clock.currentTimeNanos
+            return cacheEntries[key]?.let {
+                if (it.isExpired(nowNanos)) {
+                    // clean up expired entries
+                    expireEntries(nowNanos)
+                    null
+                } else {
+                    // update eviction metadata
+                    it.recordRead(nowNanos)
+                    it.value
+                }
+            } ?: loader().let { loadedValue ->
+                // cache the loaded value if value for the key is still absent
+                // from the cache entries after executing the loader
+                val existingValue = get(key)
+                if (existingValue != null) {
+                    existingValue
+                } else {
+                    put(key, loadedValue)
+                    loadedValue
+                }
+            }
+        }
+    }
+
+    override fun put(key: Key, value: Value) {
+        val nowNanos = clock.currentTimeNanos
+        expireEntries(nowNanos)
+
+        val existingEntry = cacheEntries[key]
+        if (existingEntry != null) {
+            // cache entry found
+            existingEntry.recordWrite(nowNanos)
+            existingEntry.value = value
```

Comment:
I think you can add `cacheEntries[key] = existingEntry` to create a memory barrier using the map access.

Reply:
any reason not to make the loader `suspend` rather than blocking? Could be follow up

Reply:
There's no `CoroutineScope` to execute the `suspend` function. If we made the `get(key, loader)` API itself `suspend` we will need to use `Mutex` which is also part of `kotlinx-coroutines-core`  for synchronization. Or do we want to introduce coroutines library dependency to `cache`?